### PR TITLE
Basic implementation of a PWA - Closes #2230

### LINKED ISSUE #2230: PWA
Basic implemention of a PWA application:

1) Installable standalone application.
2) Precache discoverable css and js.
3) network then cache pages for a better offline experience.

### PR DESCRIPTION
**Pull request by ThisUsedToBeAnEmail**
Basic implemention of a PWA application.

Installable, with an offline experience.

### CODE REVIEW

File: lib/MetaCPAN/Web/Controller/Root.pm [hunk 1]
```diff
@@ -83,6 +83,16 @@ sub robots : Path("robots.txt") : Args(0) {
     $c->stash( { template => 'robots.txt' } );
 }
 
+sub service_worker : Path("service_worker") : Args(0) {
+    my ( $self, $c ) = @_;
+    $c->res->content_type('application/javascript');
+    $c->stash( {
+        version      => 'v1.0',
```
Comment (ThisUsedToBeAnEmail) [inline lib/MetaCPAN/Web/Controller/Root.pm]: This should really come from a config file or env variable - couldn't figure out if/where a application/release version is available.

File: root/service_worker.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    "<% js %>",
+    <%- END %>
+    <%- FOREACH less IN req.env.item('psgix.assets').grep(/\.less$/) %>
+    "/static/less/style.less",
+    <%- need_less = 1 %>
+    <%- END %>
+    <%- IF need_less %>
+    "/static/js/less.min.js"
+    <%- END %>
+];
```
Comment (ThisUsedToBeAnEmail) [inline root/service_worker.js]: I'm sure I am missing some resources that can be precached.

File: root/static/manifest.json [hunk 3]
```diff
@@ -0,0 +1,99 @@
+{
+   "orientation" : "any",
+   "scope" : "/",
+   "start_url" : "/",
+   "theme_color" : "#da2037",
+   "icons" : [
+      {
+         "src" : "/static/manifest-icons/96x96-icon.png",
```
Comment (ThisUsedToBeAnEmail) [inline root/static/manifest.json]: I'm using small_logo.png but perhaps we want to use another. (just the dots but couldn't find it in the /images dir)

File: root/wrapper.html [hunk 4]
```diff
@@ -273,5 +279,42 @@ <h4 class="modal-title">Keyboard Shortcuts</h4>
             </div>
           </div>
         </div>
-    </body>
+	<script>
```
Comment (ThisUsedToBeAnEmail) [inline root/wrapper.html]: probably want to move this into it's own file /static/js

**Comment by ThisUsedToBeAnEmail** [comment]
Please do check this out and test before merging.

**Comment by mohawk2** [comment]
I believe (though I'm just an outside observer!) this would be more persuasive if a before-and-after comparison of performance were done, possibly using https://webpagetest.org/easy, and also casting an eye over https://developers.google.com/web/fundamentals/performance/rail

File: root/service_worker.js (hunk 2 above)
Comment (mohawk2) [reply root/service_worker.js:22]: The fonts?

**Comment by oalders** [comment]
@ThisUsedToBeAnEmail I'm just wondering why this was closed?

**Comment by mohawk2** [comment]
Me too. With the changes suggested by @haarg this should be mergeable, especially if a bit of performance-checking were done to make the case.

**Comment by ThisUsedToBeAnEmail** [comment]
The Idea would be to send a push notification per release, obviously it would be optional for you to subscribe to such notifications. example:  https://serviceworke.rs/push-clients_demo.html

**Comment by mohawk2** [comment]
Warning: push notifications are quite hard. My module https://metacpan.org/pod/Mojolicious::Plugin::WebPush encapsulates them (it uses my https://metacpan.org/pod/Mojolicious::Plugin::ServiceWorker because there's a lot of service-worker stuff involved).

**Pull request closed without merging**